File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        pair<int,int> mn = {INT_MAX, -1}; 
        pair<int,int> mx = {INT_MIN, -1}; 
        int total = 0; 

        for(int i = 0; i < prices.size(); i ++) {
            if(mn.first> prices[i]){
                mn = {prices[i], i};
            }

            if(prices[i] >= mx.first && i > mn.second || prices[i] - mn.first > total && i > mn.second ){
                mx = {prices[i], i}; 
                total = max(mx.first - mn.first, total); 
            }

        }
        return total; 
    }
};

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        mx = 0

        r = len(prices)-1

        gap = 0

        while r >= 0 : 
            mx = max(prices[r], mx)
            if mx > prices[r]:
                gap = max(mx - prices[r] , gap)
            r-=1

        return gap 

        


        